toolchain/semantics/semantics_handle_paren.cpp

```diff
@@ -9,27 +9,50 @@ namespace Carbon {
 auto SemanticsHandleParenExpression(SemanticsContext& context,
                                     ParseTree::Node parse_node) -> bool {
   auto value_id = context.node_stack().PopExpression();
+  // ParamOrArgStart was called for tuple handling; clean up the ParamOrArg
+  // support for non-tuple cases.
+  context.ParamOrArgEnd(
+      /*for_args=*/true, ParseNodeKind::ParenExpressionOrTupleLiteralStart);
   context.node_stack()
       .PopAndDiscardSoloParseNode<
           ParseNodeKind::ParenExpressionOrTupleLiteralStart>();
   context.node_stack().Push(parse_node, value_id);
   return true;
 }
 
-auto SemanticsHandleParenExpressionOrTupleLiteralStart(
-    SemanticsContext& context, ParseTree::Node parse_node) -> bool {
-  context.node_stack().Push(parse_node);
+auto SemanticsHandleTupleLiteralComma(SemanticsContext& context,
+                                      ParseTree::Node /*parse_node*/) -> bool {
+  context.ParamOrArgComma(/*for_args=*/true);
   return true;
 }
 
 auto SemanticsHandleTupleLiteral(SemanticsContext& context,
                                  ParseTree::Node parse_node) -> bool {
-  return context.TODO(parse_node, "HandleTupleLiteral");
+  auto refs_id = context.ParamOrArgEnd(
+      /*for_args=*/true, ParseNodeKind::ParenExpressionOrTupleLiteralStart);
+
+  context.node_stack()
+      .PopAndDiscardSoloParseNode<
+          ParseNodeKind::ParenExpressionOrTupleLiteralStart>();
+  auto node_blocks = context.semantics_ir().GetNodeBlock(refs_id);
+  llvm::SmallVector<SemanticsTypeId> type_ids;
```

Comment:
```suggestion
  llvm::SmallVector<SemanticsTypeId> type_ids;
  type_ids.reserve(node_blocks.size());
```

Reserving the size can reduce reallocations.